src/action.py

```diff
@@ -0,0 +1,344 @@
+            reason=str(exc),
+        )
+
+
+def _run_one(
+    action: types_.AnyAction,
+    discourse: Discourse,
+    draft_mode: bool,
+    delete_pages: bool,
+) -> types_.ActionReport:
+    """Take the actions against the server.
+
+    Args:
+        actions: The actions to take.
+        discourse: A client to the documentation server.
+        draft_mode: If enabled, only log the action that would be taken.
+        delete_pages: Whether to delete pages that are no longer needed.
+
+    Returns:
+        A report on the outcome of executing the action.
+
+    Raises:
+        ActionError: if an action that is not handled is passed to the function.
+    """
+    match action.action:
+        case types_.Action.CREATE:
+            # To help mypy (same for the rest of the asserts), it is ok if the assert does not run
+            assert isinstance(action, types_.CreateAction)  # nosec
+            return _create(action=action, discourse=discourse, draft_mode=draft_mode)
+        case types_.Action.NOOP:
+            assert isinstance(action, types_.NoopAction)  # nosec
+            return _noop(action=action)
+        case types_.Action.UPDATE:
+            assert isinstance(action, types_.UpdateAction)  # nosec
+            return _update(action=action, discourse=discourse, draft_mode=draft_mode)
+        case types_.Action.DELETE:
+            assert isinstance(action, types_.DeleteAction)  # nosec
+            return _delete(
+                action=action,
+                discourse=discourse,
+                draft_mode=draft_mode,
+                delete_pages=delete_pages,
+            )
+        # Edge case that should not be possible
+        case _:  # pragma: no cover
+            raise exceptions.ActionError(
+                f"internal error, no implementation for action, {action=!r}"
+            )
+
+
+def _run_index(
+    action: types_.AnyIndexAction, discourse: Discourse, draft_mode: bool
+) -> types_.ActionReport:
+    """Take the index action against the server.
+
+    Args:
+        action: The actions to take.
+        discourse: A client to the documentation server.
+        draft_mode: If enabled, only log the action that would be taken.
+
+    Returns:
+        A report on the outcome of executing the action.
+
+    Raises:
+        ActionError: if an action that is not handled is passed to the function.
+    """
+    logging.info("draft mode: %s, action: %s", draft_mode, action)
+
+    if draft_mode:
+        return types_.ActionReport(
+            table_row=None,
+            url=DRAFT_NAVLINK_LINK,
+            result=types_.ActionResult.SKIP,
+            reason=DRAFT_MODE_REASON,
+        )
+
+    match action.action:
+        case types_.Action.CREATE:
+            try:
+                # To help mypy (same for the rest of the asserts), it is ok if the assert does not
+                # run
+                assert isinstance(action, types_.CreateIndexAction)  # nosec
```

Comment:
What's the reason behind this?

Reply:
It's possible to use `typing.cast` here?

Reply:
The action that is passed to the method can be a create, update or noop action, depending on the differences between the files in the `docs` folder and the documentation on discourse. The match differentiates between the cases correctly, although mypy has difficulty interpreting that the individual `case` statements guarantee that the action is of a particular type. The asserts are there to help mypy understand that the action is guaranteed to be a particular type here. This line and the other asserts could be removed without any functional change (the tests would still pass). An alternative is to put a `type: ignore` statement, the current approach still lets future editors benefit from type hints.

Reply:
Just closing this since I believe I addressed it, please let me know if you have further feedback